File: sort/Bubble.cpp
#include <bits/stdc++.h>

using namespace std;

/* Neste exemplo, o objetivo é ordernar o array de forma crescente. 
Para que o mesmo seja decrescente, basta trocar o sinal de 'maior que' para 'menor que' */

bool need_swap(int a, int b){
	/* Condição de troca dos elementos do array */
	return a > b;
}

void swap(int * a, int * b){
	int t = *a;  /* O valor armazenado no endereço de memória representado por *a será guardado na variável t*/
	*a = *b;     
	*b = t;
}

void bubble(int array[], int array_size){
	/* O array será percorrido de dois em dois elementos */
	for(int i = 0; i < array_size-1; i++){
		for(int j = i + 1; j < array_size; j++)	{
			/* Os elementos serão analisados para saber se precisarão trocar de posição*/
			if(need_swap(array[i], array[j])){
				swap(array[i], array[j]);
			}
		}
	}
	
	return array;
}

int main(){
	/* Declarado array para ordernar */
	int array[]=  { 2, 3, 1, 5, 4 };
	/* Armazenado o tamanho do array */
	int size = sizeof(array)/sizeof(int);
	/* Chamado método de ordenamento bubblesort (método bolha) */
	bubble(array, size);

	/* Exibição do array ordenado */
	for(int a:array){
		cout << a <<endl;
	}
	
	return 0;
}

File: sort/Bubble.cs
using System;

public class BubbleSort
{
  public static void Main()
  {
    // Array de exemplo.
    int[] array = new int[] { 2, 3, 1, 5, 4 };
    
    // Resultado da ordenação do array de exemplo.
    int[] arraySorted = Bubble(array);
    
    foreach(var item in arraySorted)
    {
      Console.WriteLine(item);
    }
  }
  
  public static int[] Bubble(int[] array)
  {
    // Do primeiro ao penultimo.
    for(int a = 0; a < array.Length - 1; a++)
    {
      // A partir do "A" até o ultimo.
      for(int b = a + 1; b < array.Length; b++)
      {
        // Verificação de necessidade de troca das posição entre A e B.
        if(NeedSwap(array[a], array[b]))
        {
          Swap(array, a, b);
        }
      }
    }
    
    return array;
  }
  
  // Verifica a necessidade da troca de posição entre A e B.
  // a > b	faz com que todos da esquerda tenham que ser menor que o da direita.
  // a < b	faz com que todos da esquerda tenham que ser maior que o da direita.
  public static bool NeedSwap(int a, int b)
  {
    return a > b;
  }
  
  // Realiza a troca de posição entre A e B dentro do array.
  // aux recebe o valor de A
  // A recebe o valor de B
  // B recebe o valor de aux (valor do A antes de ser sobreescrito).
  public static void Swap(int[] array, int a, int b)
  {
    int aux = array[a];
    array[a] = array[b];
    array[b] = aux;
  }
}
